File: jvm/src/main/java/com/concord/wordcounter/WordCounter.java
package com.concord;

import com.concord.*;
import com.concord.swift.*;
import java.util.*;

public class WordCounter extends Computation {

  private final HashMap<String, Integer> histogram = new HashMap<>();
  private int pidx = 0;

  public void init(ComputationContext ctx) {
    System.out.println("WordCounter.java initialized");
  }

  public void processTimer(ComputationContext ctx, String key, long time) {
    throw new RuntimeException("Method not implemented");
  }

  public void processRecord(ComputationContext ctx, Record record) {
    try {
      String key = new String(record.getKey(), "UTF-8");
      Integer currentValue = this.histogram.get(key);
      if (currentValue != null) {
        this.histogram.put(key, ++currentValue);
      } else {
        this.histogram.put(key, 1);
      }
    } catch (Exception e) {
      throw new RuntimeException(e);
    }
    if (++this.pidx % 1024 == 0) {
      System.out.println(this.histogram.toString());
    }
  }

  public Metadata metadata() {
    Set<StreamTuple> istreams = new HashSet<StreamTuple>();
    istreams.add(new StreamTuple("words", StreamGrouping.SHUFFLE));
    return new Metadata("word-counter", istreams, new HashSet<String>());
  }

  public static void main(String[] args) {
    ServeComputation.serve(new WordCounter());
  }
}

File: scala/src/main/scala/com/concord/wordcounter/WordCounter.scala
package com.concord.wordcounter

import java.util

import com.concord._
import com.concord.swift._

class WordCounter extends Computation {

  private val histogram: util.HashMap[String, Integer] = new util.HashMap()
  private var pidx = 0

  def init(ctx: ComputationContext): Unit = {
    println(s"${this.getClass.getSimpleName} initialized")
  }

  def processTimer(ctx: ComputationContext, key: String, time: Long): Unit = {
    throw new RuntimeException("Method not implemented")
  }

  def processRecord(ctx: ComputationContext, record: Record): Unit = {
    try {
      val key = new String(record.getKey, "UTF-8")
      var currentValue = histogram.get(key)
      if (currentValue != null) {
        currentValue += 1
        histogram.put(key, currentValue)
      }
      else
        histogram.put(key, 1)
    } catch {
      case e: Exception => throw new RuntimeException(e)
    }

    pidx += 1
    if (pidx % 1024 == 0) {
      println(histogram.toString)
    }
  }

  def metadata(): Metadata = {
    val istreams = new util.HashSet[StreamTuple]()
    istreams.add(new StreamTuple("words", StreamGrouping.SHUFFLE))
    new Metadata("word-counter", istreams, new util.HashSet[String]())
  }

}


object WordCounter {
  def main(args: Array[String]): Unit = {
    ServeComputation.serve(new WordCounter())
  }
}
